### PR TITLE
[TOOLS-1894] Filtering issuesKpi by progressing timespan

### PR DESCRIPTION
**Pull request by heronoyama**

### CODE REVIEW

File: src/main/java/objective/taskboard/followup/kpi/StatusTransition.java [hunk 1]
```diff
@@ -102,6 +124,36 @@ public Long totalEffort() {
...
         return allWorklogs;
     }
+    
+    public Optional<ZonedDateTime> firstDateOnProgressing(ZoneId timezone){
+        if(isProgressingWithWorklogs())
+            return minimumDateFromWorklog(timezone);
+        return next.flatMap(n -> n.firstDateOnProgressing(timezone));
+    }
+
+    public Optional<ZonedDateTime> getDateAfterLeavingLastProgressingStatus(ZoneId timezone) {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/followup/kpi/StatusTransition.java]: MAJOR Remove this unused method parameter "timezone". [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1172)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/test/java/objective/taskboard/followup/kpi/IssueKpiServiceTest.java [hunk 2]
```diff
@@ -192,7 +184,161 @@ public void getIssues_currentState() {
...
+          
+        builder.withMockingIssue("I-2", "Alpha", SUBTASKS)
+                .setCurrentStatusToCurrentIssue("Done")
+                .setProjectKeyToCurrentIssue("PROJ")
+                .setFatherToCurrentIssue("I-1")
+                .addTransition("Open", "2020-01-01")
+                .addTransition("To Do", "2020-01-02")
+                .addTransition("Doing", "2020-01-03")
+                .addTransition("Done", "2020-01-04")
+                .addWorklog("2020-01-03",300);
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/IssueKpiServiceTest.java]:
```
builder.issue("I-2")
  .type("Alpha")
  .isSubtask()
  .withTransitions()
     .name("Open").date("2020-01-01")
     .name("To Do").date("2020-01-02")
     .name("Doing").date("2020-01-03")
     .name("Done").date("2020-01-04")
  .worklogs()
      .at("2020-01-03").timeSpent(300);
```

File: src/test/java/objective/taskboard/followup/kpi/IssueKpiServiceTest.java [hunk 3]
```diff
@@ -192,7 +184,161 @@ public void getIssues_currentState() {
...
+        
+        builder.withMockingIssue("I-3", "Alpha", SUBTASKS)
+                .setCurrentStatusToCurrentIssue("To Do")
+                .setProjectKeyToCurrentIssue("PROJ")
+                .setFatherToCurrentIssue("I-1")
+                .addTransition("Open", "2020-01-01")
+                .addTransition("To Do","2020-01-02")
+                .addTransition("Doing")
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/IssueKpiServiceTest.java]: Is a transition without a date really possible?

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    private FixedClock clock = new FixedClock();
+    private static final String START_RANGE = "2018-01-10";
+    private static final String END_RANGE = "2018-01-20";
+    private static final ZoneId timezone = ZoneId.systemDefault();
+    
+    private TouchTimeFilter subject;
+    
+    @Before
+    public void setup() {
+        clock.setNow(DateTimeUtils.parseDateTime("2018-02-25").toInstant());
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java]: could be just `clock.setNow("2018-02-25")`

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+import objective.taskboard.domain.ProjectFilterConfiguration;
+import objective.taskboard.followup.kpi.enviroment.KPIEnvironmentBuilder;
+import objective.taskboard.testUtils.FixedClock;
+import objective.taskboard.utils.DateTimeUtils;
+
+public class TouchTimeFilterTest {
+    
+    private FixedClock clock = new FixedClock();
+    private static final String START_RANGE = "2018-01-10";
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java]: Why not inline these constants?

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                .addTransition("Reviewing","2018-01-07")
+                .addTransition("Done","2018-01-10");
+        
+        IssueKpi issueKpi = builder.buildCurrentIssueAsKpi();
+        
+        assertTrue(subject.test(issueKpi));
+    }
+    
+    @Test
+    public void issueStartingAtEndOfRange() {
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java]: issueStartingAtEndOfRange_shouldBeConsideredInsideRange

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        builder.mockingSubtask("I-1", "Development")
+                .addTransition("To Do","2018-01-15")
+                .addTransition("Doing")
+                .addTransition("To Review")
+                .addTransition("Reviewing")
+                .addTransition("Done");
+        
+        IssueKpi issueKpi = builder.buildCurrentIssueAsKpi();
+        
+        assertFalse(subject.test(issueKpi));
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java]:
```
givenIssue("I-1").isSubtask()
   .transitions()
      .name("To Do").at("2018-01-15")
      .name("Doing")
   ...
.then()
   .touchTimeFilter("2018-01-10","2018-01-20").shouldRejectIssue();
   
```

File: src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        builder.mockingSubtask("I-1", "Development")
+                .addTransition("To Do","2018-01-19")
+                .addTransition("Doing","2018-01-20")
+                .addTransition("To Review","2018-01-25")
+                .addTransition("Reviewing","2018-01-27")
+                .addTransition("Done","2018-01-31");
+        
+        IssueKpi issueKpi = builder.buildCurrentIssueAsKpi();
+        
+        assertTrue(subject.test(issueKpi));
```
Comment (taksan) [inline src/test/java/objective/taskboard/followup/kpi/TouchTimeFilterTest.java]: touchTimeFilter().from("2018-01-10").to("2018-01-20").shouldAccept(issueKpi)

**Pull request merged**